File: app/src/main/java/com/dvor/my/mydvor/Storage.java
package com.dvor.my.mydvor;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.annotation.NonNull;
import android.util.Log;
import android.widget.ImageView;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.storage.FileDownloadTask;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import java.io.ByteArrayOutputStream;
import java.io.File;


public class Storage {

    public static void downloadPicture (String url, final ImageView imgView)
    {

        try {
            FirebaseStorage storage = FirebaseStorage.getInstance();
            StorageReference storageRef = storage.getReference();
            StorageReference imageRef = storageRef.child(url);
            final File localFile = File.createTempFile("image", "jpg");
            imageRef.getFile(localFile).addOnSuccessListener(new OnSuccessListener<FileDownloadTask.TaskSnapshot>() {
                @Override
                public void onSuccess(FileDownloadTask.TaskSnapshot taskSnapshot) {
                    Bitmap bitmap=BitmapFactory.decodeFile(localFile.getAbsolutePath());
                    imgView.setImageBitmap(bitmap);
                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception exception) {
                    imgView.getLayoutParams().height = 0;
                    imgView.requestLayout();
                }
            });
        }
        catch (Exception ex)
        {
            Log.d("state", ex.getMessage());
        }
    }

    public static void uploadPicture(final Bitmap img, String Url)
    {
        FirebaseStorage storage=FirebaseStorage.getInstance();
        StorageReference storageRef = storage.getReference();
        StorageReference imagesRef = storageRef.child(Url);
        try {
            MainActivity.imgPref=img;
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            img.compress(Bitmap.CompressFormat.JPEG, 100, baos);
            byte[] data = baos.toByteArray();
            UploadTask uploadTask = imagesRef.putBytes(data);

            uploadTask.addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception exception) {
                    // Handle unsuccessful uploads
                }
            }).addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
                @Override
                public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {

                }
            });
        }
        catch (Exception ex)
        {
            Log.d("state", ex.getMessage());
        }
    }

    public static void deleteImg(String imgUrl)
    {
        StorageReference storageReference = FirebaseStorage.getInstance().getReference().child(imgUrl);
        storageReference.delete().addOnSuccessListener(new OnSuccessListener<Void>() {
            @Override
            public void onSuccess(Void aVoid) {
                // File deleted successfully

            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception exception) {
                // Uh-oh, an error occurred!

            }
        });
    }
}

File: app/src/main/java/com/dvor/my/mydvor/Storage.kt
package com.dvor.my.mydvor

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import android.widget.ImageView
import com.google.firebase.storage.FirebaseStorage
import java.io.ByteArrayOutputStream
import java.io.File


object Storage {

    fun downloadPicture(context: Context, url: String, imgView: ImageView) {
        val buf = url.split('/')
        val filename = buf.last()
        val path = url.dropLast(filename.length)

        var bufPath = context.getCacheDir().toString()

        for(el in buf.dropLast(1)) {
            bufPath += "/" + el

            val f = File(bufPath)

            if(!f.exists()) {
                f.mkdir()
            }
        }

        try {
            val storage = FirebaseStorage.getInstance()
            val storageRef = storage.reference
            val imageRef = storageRef.child(url)
            val file = File(context.getCacheDir().toString() + "/" + path, filename);
            if(!file.exists()) {
                file.createNewFile()
                imageRef.getFile(file).addOnSuccessListener {
                    val bitmap = BitmapFactory.decodeFile(file.absolutePath)
                    imgView.setImageBitmap(bitmap)
                }.addOnFailureListener {
                    imgView.layoutParams.height = 0
                    imgView.requestLayout()
                }
            }
            else {
                imgView.setImageBitmap(BitmapFactory.decodeFile(file.absolutePath))
            }
        } catch (ex: Exception) {
            Log.d("state", ex.message.toString())
        }
    }

    fun uploadPicture(img: Bitmap, Url: String, f: () -> Unit) {
        val storage = FirebaseStorage.getInstance()
        val storageRef = storage.reference
        val imagesRef = storageRef.child(Url)
        try {
            MainActivity.imgPref = img
            val baos = ByteArrayOutputStream()
            img.compress(Bitmap.CompressFormat.JPEG, 100, baos)
            val data = baos.toByteArray()
            val uploadTask = imagesRef.putBytes(data)

            uploadTask.addOnSuccessListener {
                f()
            }.addOnFailureListener {
                Log.d("state", "failure upload")
            }
        } catch (ex: Exception) {
            Log.d("state", ex.message.toString())
        }

    }

    fun deleteImg(imgUrl: String) {
        val storageReference = FirebaseStorage.getInstance().reference.child(imgUrl)
        storageReference.delete().addOnSuccessListener {
            // File deleted successfully
        }.addOnFailureListener {
            // Uh-oh, an error occurred!
        }
    }
}
